scratch/copy-mk.py

```diff
@@ -185,6 +232,71 @@ def cbf_sensor_value(name: str) -> Any:
             await client.wait_closed()
+
+    if args.transfer_delays:
+        stream_names = [f"antenna-channelised-voltage-{input_label}" for input_label in input_labels]
+        # New client because it needs to be initialised with the callback (?)
+        callback_client = KATPortalClient(
+            args.portal,
+            partial(
+                delay_transfer_callback,
+                delays_queue=delays_queue,
+            ),
+        )
+        # Connect before subscribing
+        await callback_client.connect()
+        delay_sensors_regex = (
+            "^(?:"
+            + "|".join(f"{prefix}_wide_{stream_name.replace("-", "_")}_delay" for stream_name in stream_names)
+            + ")$"
+        )
+        # TODO: Check against results returned from the following to
+        # ensure the `callback_client` is ready to proceed.
+        namespace = f"{args.name}_{uuid.uuid4()}"
+        await callback_client.subscribe(namespace)
+        await callback_client.set_sampling_strategies(namespace, delay_sensors_regex, "event")
+
+        def _create_empty_dict(keys: list[Any]) -> dict[Any, None]:
+            empty_dict = {new_key: None for new_key in keys}
+            return empty_dict
+
+        delay_args: dict[str, tuple[int, str, str] | None] = _create_empty_dict(input_labels)
+        ref_loadmcnt = 0
+        try:
+            pc_client = await aiokatcp.Client.connect(pc_host, pc_port)
+            while True:
+                complete_requests = False
+                while not complete_requests:
+                    # delays_queue items are in the format
+                    # - (loadmcnt, src_name, delay_request_args)
+                    delays_queue_item = await delays_queue.get()
+                    if ref_loadmcnt == 0:
+                        # This is the first delay update we're seeing
+                        # TODO: Potentially move this back outside the while
+                        # so it doesn't need to be checked on every loop.
+                        ref_loadmcnt = delays_queue_item[0]
+                    if delays_queue_item[0] != ref_loadmcnt:
+                        ref_loadmcnt = delays_queue_item[0]
+                        print(f"{delays_queue_item[1]}: New loadmcnt {ref_loadmcnt}. Getting new delays.")
+                        delay_args = _create_empty_dict(input_labels)
+
+                    delay_args[delays_queue_item[1]] = delays_queue_item
+                    complete_requests = all(delay_args_value is not None for delay_args_value in delay_args.values())
+
+                unix_timestamp = ref_loadmcnt / scale_factor_timestamp + sync_time
+                sorted_delay_args = dict(sorted(delay_args.items()))
+                await pc_client.request(
+                    "delays",
+                    "antenna-channelised-voltage",
+                    unix_timestamp,
+                    *[delay_arg_value[-1] for delay_arg_value in sorted_delay_args.values()],
```

Comment:
This bit was a bit awkward to navigate. I could've done the `sorted(...)` in-line but the list comprehension wouldn't be that easy to comprehend. Happy to change if needed.